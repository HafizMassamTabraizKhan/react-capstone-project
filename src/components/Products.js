import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { BsArrowRightCircle } from 'react-icons/bs';
import { Link } from 'react-router-dom';
import { fetchProducts } from '../redux/products/productsSlice';

function Products() {
  const dispatch = useDispatch();
  const products = useSelector((state) => state.products.products);
  const status = useSelector((state) => state.products.status);
  const error = useSelector((state) => state.products.error);
  const [search, setSearch] = useState('');

  useEffect(() => {
    dispatch(fetchProducts());
  }, [dispatch]);

  if (status === 'loading') {
    return <div>Loading...</div>;
  }

  if (status === 'failed') {
    return (
      <div>
        Error:
        {error}
      </div>
    );
  }

  return (
    <>
      <div className="search">
        <input
          className="input"
          type="text"
          onChange={(e) => setSearch(e.target.value)}
          placeholder="Search by category"
          value={search}
        />
      </div>
      <div className="products-container">
        <div className="products-grid">
          {products
            .filter((product) => {
              if (search === '') {
                return product;
              }
              if (product.category.toLowerCase().includes(search.toLowerCase())) {
                return product;
              }
              return null;
            })
            .map((product, index) => (
              <div className={`products-item ${index % 2 === 0 ? 'background-1' : 'background-2'}`} key={product.id}>
                <img src={product.image} alt="" />
                <h3>{product.title}</h3>
                <div className="price-category">
                  <p>{product.category}</p>
                  <p>
                    $
                    {product.price}
                  </p>
                </div>
                <div className="view-more">
                  <span>Click for more info</span>
                  <Link to={`/product/${product.id}`}><BsArrowRightCircle /></Link>
                </div>
              </div>
            ))}
        </div>
      </div>
    </>
  );
}

export default Products;
