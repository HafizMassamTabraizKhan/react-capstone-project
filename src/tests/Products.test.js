import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import '@testing-library/jest-dom';
import store from '../redux/store';
import Products from '../components/Products';

test('Check any changes to the component', () => {
  const tree = renderer.create(
    <Provider store={store}>
      <Products />
    </Provider>,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

it('the component should render a div container', () => {
  const { getByTestId } = render(
    <Provider store={store}>
      <Products />
    </Provider>,
  );
  const container = getByTestId('productsPage');
  expect(container).toBeInTheDocument();
});
