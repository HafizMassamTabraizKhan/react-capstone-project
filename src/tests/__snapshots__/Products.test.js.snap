// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check any changes to the component 1`] = `
<div
  data-testid="productsPage"
>
  <div
    className="search"
  >
    <input
      className="input"
      onChange={[Function]}
      placeholder="Search by category"
      type="text"
      value=""
    />
  </div>
  <div
    className="products-container"
  >
    <div
      className="products-grid"
    />
  </div>
</div>
`;
